search/streets_matcher.cpp

```diff
@@ -44,10 +79,25 @@ void StreetsMatcher::Go(BaseContext const & ctx, FeaturesFilter const & filter,
   if (predictions.empty())
     return;
 
-  sort(predictions.begin(), predictions.end(), &LessByHash);
-  predictions.erase(
-      unique(predictions.begin(), predictions.end(), base::EqualsBy(&Prediction::m_hash)),
-      predictions.end());
+  // Remove predictions with the same m_hash (features) and token range.
+  sort(predictions.begin(), predictions.end(), &LessByHashAndRange);
+  predictions.erase(unique(predictions.begin(), predictions.end(), &EqualsByHashAndRange),
+                    predictions.end());
+
+  // Leave the most probable and longest prediction for predictions with the same m_hash (features)
+  // and m_withMisprints.
+  // We will still distinguish parses with the same m_hash (features) but different range and m_withMisprints.
+  // For example, for "Paramount dive" we will have two parses:
+  // STREET      UNUSED (can be matched to poi later)
+  // Paramount   dive
+  //
+  // STREET       STREET ("drive" with misprints)
+  // Paramount    dive
+  //
+  // Both parses will have the same features and hash but we need both of them.
```

Comment:
s/need both of them/cannot discard either of them/
or
s/but/yet/
or
just replace the first "Both" with "The"

Reply:
done